File: rust/src/day21.rs
use std::collections::HashMap;

#[derive(Eq, Hash, PartialEq)]
struct Universe {
    positions: (u8, u8),
    scores: (u8, u8),
}

impl Universe {
    pub fn new(positions: (u8, u8), scores: (u8, u8)) -> Self {
        Universe { positions, scores }
    }
}

pub struct Submarine {
    positions: (u8, u8),
    scores: (u32, u32),
    die_rolls: u32,
    universe_scores: (u64, u64),
}

impl Submarine {
    pub fn new() -> Self {
        Self {
            positions: (0, 0),
            scores: (0, 0),
            die_rolls: 0,
            universe_scores: (0, 0),
        }
    }

    pub fn on_input(&mut self, result: &str) {
        if self.positions.0 == 0 {
            self.positions.0 = result.strip_prefix("Player 1 starting position: ").unwrap().parse::<u8>().unwrap() - 1;
        } else {
            self.positions.1 = result.strip_prefix("Player 2 starting position: ").unwrap().parse::<u8>().unwrap() - 1;
        }
    }

    // Returns the number of positions to move forward after all the complete rotations around the board.
    fn roll_d100(&mut self) -> u8 {
        let res = 1 + (self.die_rolls % 100);
        self.die_rolls += 1;
        (res % 10) as u8
    }

    fn run_practice_game(&mut self) {
        let mut game_positions = self.positions;

        loop {
            let p1_dice = self.roll_d100() + self.roll_d100() + self.roll_d100();
            game_positions.0 = (game_positions.0 + p1_dice) % 10;
            self.scores.0 += (1 + game_positions.0) as u32;

            if self.scores.0 >= 1000 {
                return;
            }

            let p2_dice = self.roll_d100() + self.roll_d100() + self.roll_d100();
            game_positions.1 = (game_positions.1 + p2_dice) % 10;
            self.scores.1 += (1 + game_positions.1) as u32;

            if self.scores.1 >= 1000 {
                return;
            }
        }
    }

    fn run_dirac_game(&mut self) {
        let dirac_possibilities = generate_dirac_universes(3);

        let mut universes: HashMap<Universe, u64> = HashMap::new();
        let starting_universe = Universe::new(self.positions, (0, 0));
        universes.insert(starting_universe, 1);

        loop {
            if universes.len() > 500_000 {
                println!("WARNING");
                return;
            }

            let mut new_universes: HashMap<Universe, u64> = HashMap::new();

            for (dice_sum, possibilities) in dirac_possibilities.iter().enumerate() {
                for (universe, amount) in universes.iter() {
                    let p1_location = (universe.positions.0 + (dice_sum as u8) + 3) % 10;
                    let p1_score = universe.scores.0 + p1_location + 1;

                    if p1_score >= 21 {
                        self.universe_scores.0 += amount * possibilities;
                        continue;
                    }

                    let new_universe = Universe::new((p1_location, universe.positions.1), (p1_score, universe.scores.1));
                    *new_universes.entry(new_universe).or_default() += amount * possibilities;
                }
            }

            if new_universes.len() == 0 {
                return;
            }

            universes = new_universes;

            new_universes = HashMap::new();

            for (dice_sum, possibilities) in dirac_possibilities.iter().enumerate() {
                for (universe, amount) in universes.iter() {
                    let p2_location = (universe.positions.1 + (dice_sum as u8) + 3) % 10;
                    let p2_score = universe.scores.1 + p2_location + 1;

                    if p2_score >= 21 {
                        self.universe_scores.1 += amount * possibilities;
                        continue;
                    }

                    let new_universe = Universe::new((universe.positions.0, p2_location), (universe.scores.0, p2_score));
                    *new_universes.entry(new_universe).or_default() += amount * possibilities;
                }
            }

            if new_universes.len() == 0 {
                return;
            }

            universes = new_universes;
        }
    }

    pub fn output(&mut self) {
        self.run_practice_game();

        let loser_score = self.scores.0.min(self.scores.1);

        println!("Part 1: {}", loser_score * self.die_rolls);

        self.run_dirac_game();

        let winner_universes = self.universe_scores.0.max(self.universe_scores.1);

        println!("Part 2: {}", winner_universes);
    }
}

fn generate_dirac_universes(num_rolls: usize) -> Vec<u64> {
    let total_sum_possibilities = num_rolls * 2 + 1;
    let mut result = vec![0; total_sum_possibilities];

    add_dirac_possibility(num_rolls as u64, 3, 0, &mut result);

    result
}

fn add_dirac_possibility(min_score: u64, rolls_remaining: usize, curr_score: u64, possibilities: &mut Vec<u64>) {
    if rolls_remaining == 0 {
        possibilities[(curr_score - min_score) as usize] += 1;
        return;
    }

    add_dirac_possibility(min_score, rolls_remaining - 1, curr_score + 1, possibilities);
    add_dirac_possibility(min_score, rolls_remaining - 1, curr_score + 2, possibilities);
    add_dirac_possibility(min_score, rolls_remaining - 1, curr_score + 3, possibilities);
}

File: 2021/julia/Day21.jl
module Day21

export Submarine, on_input!, output

mutable struct Universe
    positions::Vector{UInt8}
    scores::Vector{UInt16}
    next_player::UInt8

    Universe(positions) = new(positions, [0, 0], 1)
end

mutable struct Submarine
    starting_positions::Vector{UInt8}
    d100_rolls::UInt32
    universe_scores::Vector{UInt64}

    Submarine() = new([0, 0], 0, [0, 0])
end

function on_input!(s::Submarine, a::AbstractString)
    if s.starting_positions[1] == 0
        s.starting_positions[1] = parse(UInt8, chop(a, head = length("Player 1 starting position: "), tail = 0)) - 1
    else
        s.starting_positions[2] = parse(UInt8, chop(a, head = length("Player 2 starting position: "), tail = 0)) - 1
    end
end

function d100_sum!(s::Submarine)::UInt8
    res::UInt8 = (6 + 3 * s.d100_rolls) % 100
    s.d100_rolls += 3
    res
end

function practice_game!(s::Submarine)::Universe
    u = Universe(copy(s.starting_positions))

    while true
        u.positions[u.next_player] += d100_sum!(s)
        u.positions[u.next_player] %= 10
        u.scores[u.next_player] += 1 + u.positions[u.next_player]

        if u.scores[u.next_player] >= 1000
            return u
        end

        u.next_player = u.next_player == 1 ? 2 : 1
    end
end

function dirac_game!(s::Submarine)
    dirac_possibilities = dirac_universes(3 % UInt)

    universes = Dict{NTuple{5,UInt},UInt64}()
    p1, p2 = s.starting_positions
    universes[(p1, 0, p2, 0, 1)] = 1

    while true
        new_universes = Dict{NTuple{5,UInt},UInt64}()

        for (dice_sum, possibilities) in enumerate(dirac_possibilities)
            for (u, amount) in universes
                p1, p1_s, p2, p2_s, np = u

                curr_pos = np == 1 ? p1 : p2
                curr_score = np == 1 ? p1_s : p2_s

                new_position = (curr_pos + dice_sum + 2) % 10
                new_score = curr_score + new_position + 1

                if new_score >= 21
                    s.universe_scores[np] += amount * possibilities
                    continue
                end

                if np == 1
                    p1 = new_position
                    p1_s = new_score
                    np = 2
                else
                    p2 = new_position
                    p2_s = new_score
                    np = 1
                end

                new_u = (p1, p1_s, p2, p2_s, np)

                if haskey(new_universes, new_u)
                    new_universes[new_u] += amount * possibilities
                else
                    new_universes[new_u] = amount * possibilities
                end
            end
        end

        if isempty(new_universes)
            return
        end

        universes = new_universes
    end
end

function output(s::Submarine)
    practice_universe = practice_game!(s)

    loser_score = minimum(practice_universe.scores)

    println("Part 1: $(loser_score * s.d100_rolls)")

    dirac_game!(s)

    println("Part 2: $(maximum(s.universe_scores))")
end

function dirac_universes(num_rolls::UInt)::Vector{UInt64}
    total_sum_possibilities = num_rolls * 2 + 1
    result = zeros(UInt64, total_sum_possibilities)

    add_dirac_possibility!(result, num_rolls, num_rolls, 0%UInt64)

    result
end

function add_dirac_possibility!(result::Vector{UInt64}, min_score::UInt, rolls_remaining::UInt, curr_score::UInt64)
    if rolls_remaining == 0
        result[curr_score-min_score+1] += 1
        return
    end

    add_dirac_possibility!(result, min_score, rolls_remaining - 1, curr_score + 1)
    add_dirac_possibility!(result, min_score, rolls_remaining - 1, curr_score + 2)
    add_dirac_possibility!(result, min_score, rolls_remaining - 1, curr_score + 3)
end

end